oeps/oep-0016.rst

```diff
@@ -0,0 +1,261 @@
+open source project, it is straightforward to track the work, and to provide
+fixes as necessary.
+
+Right-to-Left Language Support
+------------------------------
+
+Bootstrap 4 does not currently provide support for right-to-left languages,
+which is clearly a roadblock for Open edX to adopt it. The official stance
+from Bootstrap is that this support will be added in a minor release after
+Bootstrap 4.0 is released. Given this, a solution needs to be found if
+Open edX is to adopt Bootstrap earlier.
+
+There were two primary options considered:
+
+* Use `bi-app-sass`_ as this is the current technology used by Open edX. This
+  approach generates two CSS files, one each for LTR and RTL.
+* Use `rtlcss`_ which is a CSS postprocessor plugin for `postcss`_. This
+  approach dynamically adds RTL-specific rules into the CSS file so that it
+  can handle both directions.
+
+The `Open edX Front End Working Group`_ discussed both options and decided that
+there are several benefits of `rtlcss` over bi-app-sass:
+
+* The default is for right/left properties to be switched which is usually
+  the desired result. With bi-app-sass, every property that needs to switch
+  needs to use a function, and it is very easy to forget to do so.
+* For special case situations, rtlcss provides declarative comments that give
+  the developer to control the behavior.
+* Having only a single CSS file that adapts to the desired language
+  direction makes asset management simpler.
+
+Given this decision, the reference implementation was updated to use `rtlcss`.
+This proof-of-concept successfully demonstrated pages rendered right-to-left
+which confirmed the choice.
+
+Note: once Bootstrap chooses its own approach, it will be necessary to revisit
+this decision. It might be simpler, for example, to switch to use the same
+technology for simplicities sake.
+
+For more background, see the `Bootstrap RTL Discovery ticket`_.
+
+Theming and Multi-Tenancy
+-------------------------
+
+A key question for Bootstrap adoption is how to implement theming for an
+application that supports multi-tenancy. Full theming encompasses many
+dimensions such as CSS, images, templates, and even feature configuration. For
+the purposes of an Open edX Bootstrap implementation, a solution needs to
+compile new versions of each application CSS file that includes the tenant's
+Bootstrap theme partial file. The application must then return the tenant's
+version of each CSS file when rendered.
+
+The technical approach for multi-tenancy is out-of-scope of this OEP, but
+it is important that the edx-bootstrap library take this requirement into
+account.
+
+Backward Compatibility
+======================
+
+For edx-platform, the approach taken in the reference implementation was to
+introduce a new v3 style, thus leaving all v1 and v2 styles unaffected. This
+allows pages to be converted one at a time.
+
+An investigation was performed to see whether Bootstrap components could be used
```

Comment:
The approach I took in the reference implementation was to conditionalize the header and footer templates with a "use_bootstrap" context parameter. If Bootstrap is being used, the template outputs Bootstrap components. I'll add this explanation.